apps/src/p5lab/spritelab/commands/criterionCommands.js

```diff
@@ -232,141 +212,141 @@ export const commands = {
   // Returns true if any sprite's costume changed this frame.
   anyCostumeChanged() {
     const spriteIds = this.getSpriteIdsInUse();
-    let result = false;
-    for (let i = 0; i < spriteIds.length; i++) {
-      const currentCostume =
-        this.nativeSpriteMap[spriteIds[i]].getAnimationLabel();
+    for (const spriteId of spriteIds) {
+      const currentCostume = this.nativeSpriteMap[spriteId].getAnimationLabel();
       const previousCostume =
-        this.previous.sprites === undefined
-          ? currentCostume
-          : this.previous.sprites.find(sprite => sprite.id === spriteIds[i])
-              .costume;
+        this.previous.sprites?.find(sprite => sprite.id === spriteId)
+          ?.costume ?? currentCostume;
+
       if (currentCostume !== previousCostume) {
-        result = true;
+        return true;
       }
     }
-    return result;
+    return false;
   },
 
   // Returns true if sprite costumes change, but only event sprites.
   // Returns false if non-event sprites change costume, or if no sprites change costume.
   onlyClickedCostumeChanged() {
     const spriteIds = this.getSpriteIdsInUse();
     const eventSpriteIds = validationCommands.getEventSpriteIds.call(this);
-    let result = false;
     let foundEventSpriteChange = false;
     let foundNoneventSpriteChange = false;
-    for (let i = 0; i < spriteIds.length; i++) {
-      const currentCostume =
-        this.nativeSpriteMap[spriteIds[i]].getAnimationLabel();
+
+    for (const spriteId of spriteIds) {
+      const currentCostume = this.nativeSpriteMap[spriteId].getAnimationLabel();
       const previousCostume =
-        this.previous.sprites === undefined
-          ? currentCostume
-          : this.previous.sprites.find(sprite => sprite.id === spriteIds[i])
-              .costume;
-      if (
-        currentCostume !== previousCostume &&
-        eventSpriteIds.includes(spriteIds[i])
-      ) {
-        foundEventSpriteChange = true;
-      } else if (currentCostume !== previousCostume) {
-        foundNoneventSpriteChange = true;
+        this.previous.sprites?.find(sprite => sprite.id === spriteId)
+          ?.costume ?? currentCostume;
+
+      if (currentCostume !== previousCostume) {
+        if (eventSpriteIds.includes(spriteId)) {
+          foundEventSpriteChange = true;
+        } else {
+          foundNoneventSpriteChange = true;
+        }
+
+        if (foundEventSpriteChange && foundNoneventSpriteChange) {
+          return false; // Found both event and non-event sprite changes, no need to continue
+        }
       }
     }
-    result = foundEventSpriteChange && !foundNoneventSpriteChange;
-    return result;
+
+    return foundEventSpriteChange && !foundNoneventSpriteChange;
```

Comment:
And then here, I think you could just
```
return foundEventSpriteChange;
```

But I think I've got too sucked into reading these validation functions and this might be too granular 😅